src/Mvc/Mvc.ViewFeatures/src/Infrastructure/HttpUriHelper.cs

```diff
@@ -0,0 +1,80 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.AspNetCore.Components.Services;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Http.Extensions;
+
+namespace Microsoft.AspNetCore.Mvc.ViewFeatures
+{
+    internal class HttpUriHelper : UriHelperBase
+    {
+        private HttpContext _context;
+
+        public HttpUriHelper()
+        {
+        }
+
+        public void InitializeState(HttpContext context)
+        {
+            _context = context;
+            InitializeState();
+        }
+
+        protected override void InitializeState()
+        {
+            if (_context == null)
+            {
+                throw new InvalidOperationException($"'{typeof(HttpUriHelper)}' not initialized.");
+            }
+            SetAbsoluteBaseUri(GetContextBaseUri());
+            SetAbsoluteUri(GetFullUri());
+        }
+
+        private string GetFullUri()
+        {
+            var request = _context.Request;
+            return UriHelper.BuildAbsolute(
+                request.Scheme,
+                request.Host,
+                request.PathBase,
+                request.Path,
+                request.QueryString);
+        }
+
+        private string GetContextBaseUri()
+        {
+            var request = _context.Request;
+            return UriHelper.BuildAbsolute(request.Scheme, request.Host, request.PathBase);
+        }
+
+        protected override void NavigateToCore(string uri, bool forceLoad)
+        {
+            // For now throw.
+            // We don't have a good way to do redirects on a prerrendering context.
+            // Currently there's no good way to stop the rendering process. We can fix
+            // this by passing in a cancellation token to RenderRootComponentAsync and
+            // having that checked before we await on every ProcessAsynchronousWork iteration
+            // if we choose to go that way. Upon redirect we could trigger the cancellation here
+            // and do _context.Response.Redirect(uri)
+            // The bigger (not easily solvable problem) is that the response might have already
+            // started (for example when the page/view has already been flushed) which will mean
+            // that the behavior of this feature is not reliable, and not easily to diagnose as
+            // the developer only notices it when this happens at runtime; something we don't really want.
+            throw new InvalidOperationException(
+                "Redirects are not supported on a prerrendering environment.");
+
+            // We need to do something here about the hash to prevent infinite redirects
+            // as the hash portion of a url doesn't get sent to the browser.
```

Comment:
Ah, I see this has been migrated forwards from an earlier PR. As per comments on the other PR, I don't follow why the hash part of the URL is a relevant thing for us to be concerned about:

 * It does get sent to the browser (it's the *server* that it doesn't get sent to), so the comment on L69 is backwards
 * We don't use hash-based routing in the template, nor do we recommend anyone does so. So I'm not sure why there'd be a risk of "infinite redirects" any more than risk of any other bug to do with someone not understanding that the server doesn't see that data

So, this is not to say anything in the code should change, but rather maybe we should drop this comment block because it could be misleading.

Reply:
A SPA oauth callback could be a place where this could cause issues depending on the implementation. But let's table this.

Reply:
I just got rid of this comment.